resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -216,8 +216,70 @@ private[spark] class Client(
                 .endContainer()
               .endSpec()
             .done()
-          submitCompletedFuture.get(30, TimeUnit.SECONDS)
-        }
+          try {
+            submitCompletedFuture.get(30, TimeUnit.SECONDS)
```

Comment:
I was fairly certain that in YARN at least the application remains in "ACCEPTED" state for some time and then spark-submit gives up running the application after some time. It could be the case that different cluster managers have different behavior though, which would mean we would have to decide on one or the other for this one.

The problem is that it's difficult to differentiate between queued versus image pull errors. Is it possible for K8s to change the "image pull failed" case to be one that's a `FAILED` state rather than a `PENDING` state? Are there any other error cases where a `PENDING` state doesn't correspond to  resource constraints? I was actually pretty surprised that the pod's state was stuck in `PENDING` when it seemed to be the result of an error to fetch the image.

If we could differentiate `PENDING` from `FAILED` and confidently say that the `PENDING` state is always one where the pod could eventually be able to run (e.g. after resources are freed up), then we could properly abort if we see the container hit the `FAILED` state, and keep SparkSubmit around in `PENDING` states. The thing I want to avoid here is keeping spark-submits running against things that have zero chance to start running successfully to begin with, like error cases. I'm fine with stalling and keeping the SparkSubmit process running to wait for resource availability.